planner/core/planbuilder.go

```diff
@@ -601,25 +607,66 @@ func (b *PlanBuilder) popSelectOffset() {
 	b.selectOffset = b.selectOffset[:len(b.selectOffset)-1]
 }
 
-// NewPlanBuilder creates a new PlanBuilder. Return the original PlannerSelectBlockAsName as well, callers decide if
+// NewPlanBuilder creates a new PlanBuilder.
+func NewPlanBuilder() *PlanBuilder {
+	return &PlanBuilder{
+		outerCTEs:           make([]*cteInfo, 0),
+		colMapper:           make(map[*ast.ColumnNameExpr]int),
+		handleHelper:        &handleColHelper{id2HandleMapStack: make([]map[int64][]HandleCols, 0)},
+		correlatedAggMapper: make(map[*ast.AggregateFuncExpr]*expression.CorrelatedColumn),
+	}
+}
+
+// Init initialize a PlanBuilder.
+// Return the original PlannerSelectBlockAsName as well, callers decide if
 // PlannerSelectBlockAsName should be restored after using this builder.
-func NewPlanBuilder(sctx sessionctx.Context, is infoschema.InfoSchema, processor *hint.BlockHintProcessor) (*PlanBuilder, []ast.HintTable) {
+// This is The comman code pattern to use it:
+// NewPlanBuilder().Init(sctx, is, processor)
+func (b *PlanBuilder) Init(sctx sessionctx.Context, is infoschema.InfoSchema, processor *hint.BlockHintProcessor) (*PlanBuilder, []ast.HintTable) {
 	savedBlockNames := sctx.GetSessionVars().PlannerSelectBlockAsName
 	if processor == nil {
 		sctx.GetSessionVars().PlannerSelectBlockAsName = nil
 	} else {
 		sctx.GetSessionVars().PlannerSelectBlockAsName = make([]ast.HintTable, processor.MaxSelectStmtOffset()+1)
 	}
-	return &PlanBuilder{
-		ctx:                 sctx,
-		is:                  is,
-		outerCTEs:           make([]*cteInfo, 0),
-		colMapper:           make(map[*ast.ColumnNameExpr]int),
-		handleHelper:        &handleColHelper{id2HandleMapStack: make([]map[int64][]HandleCols, 0)},
-		hintProcessor:       processor,
-		correlatedAggMapper: make(map[*ast.AggregateFuncExpr]*expression.CorrelatedColumn),
-		isForUpdateRead:     sctx.GetSessionVars().IsPessimisticReadConsistency(),
-	}, savedBlockNames
+
+	b.ctx = sctx
+	b.is = is
+	b.hintProcessor = processor
+	b.isForUpdateRead = sctx.GetSessionVars().IsPessimisticReadConsistency()
+	return b, savedBlockNames
+}
+
+// ResetForReuse reset the plan builder, put it into pool for reuse.
+// After reset for reuse, the object should be equal to a object returned by NewPlanBuilder().
+func (b *PlanBuilder) ResetForReuse() *PlanBuilder {
+	// Save some fields for reuse.
+	saveOuterCTEs := b.outerCTEs[:0]
+	saveColMapper := b.colMapper
+	for k := range saveColMapper {
+		delete(saveColMapper, k)
```

Comment:
Why we need to delete those keys here? Is there any difference  between deleting keys and remake a new map in this case?

Reply:
Allocation is expensive, especially for a GC language like Go.

```
for k := range m {
        delete(m, k)
}
```

is optimized to a function `runtime.mapclear()`,  while make(map[]) is also a function call.
The they're both one function call, the different is, one spent time on allocation, the other spent time on clear up the map.

So here is a tiny benchmark to compare:

```
package main

import (
	"testing"
)

func BenchmarkReuse(b *testing.B) {
	m := make(map[int]interface{})
	b.ResetTimer()
	for i:=0; i<b.N; i++ {
		use(m)
		for k := range m {
			delete(m, k)
		}
	}
}

func BenchmarkDrop(b *testing.B) {
	var m map[int]interface{}
	b.ResetTimer()
	for i:=0; i<b.N; i++ {
		m = make(map[int]interface{})
		use(m)
	}
}

func use(m map[int]interface{}) {
	m[32] = "23"
	m[46] = "421"
	m[23] = "423"
	m[52] = "666"
	m[52] = "54"
}
```

The result is:

```
genius@genius-System-Product-Name:~/project/sss$ go test -run XXX  -bench Benchmark -benchmem
goos: linux
goarch: amd64
pkg: sss
cpu: AMD Ryzen 7 PRO 4750G with Radeon Graphics
BenchmarkReuse-16    	16455187	        72.58 ns/op	       0 B/op	       0 allocs/op
BenchmarkDrop-16     	 6761996	       178.3 ns/op	     256 B/op	       2 allocs/op
PASS
ok  	sss	2.657s
```

As you can see the result ---- "allocation is expensive"

Reply:
Thanks!